File: strands-py/src/strands/agent/conversation_manager/compression/context_compression.py
"""Shared helpers for context compression strategies.

These functions are used by both the conversation managers (reactive/proactive
compression) and the agentic context-management tools (model-driven compression).
They cover the low-level mechanics: finding safe split/trim boundaries that don't
break tool-use/tool-result pairs, generating a summary via the model, and filtering
messages by type.
"""

import logging
from typing import TYPE_CHECKING, Literal, cast

from ....event_loop.streaming import process_stream
from ....types.content import Message
from ....types.exceptions import ContextWindowOverflowException

if TYPE_CHECKING:
    from ....models.model import Model

logger = logging.getLogger(__name__)


DEFAULT_SUMMARIZATION_PROMPT = """You are a conversation summarizer. Provide a concise summary of the conversation \
history.

Format Requirements:
- You MUST create a structured and concise summary in bullet-point format.
- You MUST NOT respond conversationally.
- You MUST NOT address the user directly.
- You MUST NOT comment on tool availability.

Assumptions:
- You MUST NOT assume tool executions failed unless otherwise stated.

Task:
Your task is to create a structured summary document:
- It MUST contain bullet points with key topics and questions covered
- It MUST contain bullet points for all significant tools executed and their results
- It MUST contain bullet points for any code or technical information shared
- It MUST contain a section of key insights gained
- It MUST format the summary in the third person

Example format:

## Conversation Summary
* Topic 1: Key information
* Topic 2: Key information

## Tools Executed
* Tool X: Result Y"""


MessageType = Literal["tools", "messages", "all"]
"""Filter selecting which messages a compression operation targets.

- ``"tools"``: only messages containing a toolUse or toolResult block.
- ``"messages"``: only messages without any toolUse or toolResult block.
- ``"all"``: every message.
"""


def adjust_split_point_for_tool_pairs(messages: list[Message], split_point: int) -> int:
    """Adjust a split point forward to avoid breaking toolUse/toolResult pairs.

    Walks the split point forward until the message at that position is neither an
    orphaned toolResult nor a toolUse without an immediately following toolResult.

    Args:
        messages: The full list of messages.
        split_point: The initially calculated split point.

    Returns:
        The adjusted split point that doesn't break a toolUse/toolResult pair.

    Raises:
        ContextWindowOverflowException: If the split point exceeds the message array length,
            or if no valid split point can be found (walked past all messages).
    """
    if split_point > len(messages):
        raise ContextWindowOverflowException("Split point exceeds message array length")

    if split_point == len(messages):
        return split_point

    # Find the next valid split point
    while split_point < len(messages):
        if (
            # Oldest message cannot be a toolResult because it needs a toolUse preceding it
            any("toolResult" in content for content in messages[split_point]["content"])
            or (
                # Oldest message can be a toolUse only if a toolResult immediately follows it.
                any("toolUse" in content for content in messages[split_point]["content"])
                and split_point + 1 < len(messages)
                and not any("toolResult" in content for content in messages[split_point + 1]["content"])
            )
        ):
            split_point += 1
        else:
            break
    else:
        # If we didn't find a valid split point, then we throw
        raise ContextWindowOverflowException("Unable to trim conversation context!")

    return split_point


def find_valid_trim_point(messages: list[Message], start_index: int) -> int:
    """Find a valid trim point for truncation starting at ``start_index``.

    A valid trim point must:

    1. Be a user message (required by most model providers)
    2. Not be an orphaned toolResult
    3. Not be a toolUse unless its toolResult immediately follows

    Args:
        messages: The full list of messages.
        start_index: The index to begin searching from.

    Returns:
        The valid trim index, or ``len(messages)`` if none is found.
    """
    trim_index = start_index

    while trim_index < len(messages):
        message = messages[trim_index]

        if message["role"] != "user":
            trim_index += 1
            continue

        if any("toolResult" in content for content in message["content"]):
            trim_index += 1
            continue

        if any("toolUse" in content for content in message["content"]):
            next_has_tool_result = trim_index + 1 < len(messages) and any(
                "toolResult" in content for content in messages[trim_index + 1]["content"]
            )
            if not next_has_tool_result:
                trim_index += 1
                continue

        break

    return trim_index


async def generate_summary(
    messages_to_summarize: list[Message],
    model: "Model",
    system_prompt: str | None = None,
) -> Message:
    """Generate a summary of the provided messages by calling the model directly.

    This bypasses the full agent pipeline (lock, metrics, traces, tool loop) and simply
    asks the underlying model to summarize the conversation.

    Args:
        messages_to_summarize: The messages to summarize.
        model: The model used to generate the summary.
        system_prompt: Optional system prompt override. Defaults to
            :data:`DEFAULT_SUMMARIZATION_PROMPT`.

    Returns:
        A user-role message containing the model-generated summary.

    Raises:
        RuntimeError: If the model fails to produce a response.
    """
    resolved_system_prompt = system_prompt if system_prompt is not None else DEFAULT_SUMMARIZATION_PROMPT

    summarization_messages = list(messages_to_summarize) + [
        {"role": "user", "content": [{"text": "Please summarize this conversation."}]}
    ]

    chunks = model.stream(
        summarization_messages,
        tool_specs=None,
        system_prompt=resolved_system_prompt,
    )

    result_message: Message | None = None
    async for event in process_stream(chunks):
        if "stop" in event:
            _, result_message, _, _ = event["stop"]

    if result_message is None:
        raise RuntimeError("Failed to generate summary: no response from model")

    # Return the summary as a user-role message so it's valid as conversation history
    return cast(Message, {**result_message, "role": "user"})


def matches_message_type(message: Message, filter: MessageType) -> bool:
    """Return True if the message matches the given type filter.

    Args:
        message: The message to test.
        filter: The message-type filter (``"tools"``, ``"messages"``, or ``"all"``).

    Returns:
        True if the message matches the filter.
    """
    if filter == "all":
        return True
    has_tool = any("toolUse" in content or "toolResult" in content for content in message.get("content", []))
    if filter == "tools":
        return has_tool
    if filter == "messages":
        return not has_tool
    return False  # type: ignore[unreachable]

File: strands-ts/src/conversation-manager/compression/context-compression.ts
import { Message, TextBlock } from '../../types/messages.js'
import type { Model } from '../../models/model.js'

export const DEFAULT_SUMMARIZATION_PROMPT = `You are a conversation summarizer. Provide a concise summary of the conversation \
history.

Format Requirements:
- You MUST create a structured and concise summary in bullet-point format.
- You MUST NOT respond conversationally.
- You MUST NOT address the user directly.
- You MUST NOT comment on tool availability.

Assumptions:
- You MUST NOT assume tool executions failed unless otherwise stated.

Task:
Your task is to create a structured summary document:
- It MUST contain bullet points with key topics and questions covered
- It MUST contain bullet points for all significant tools executed and their results
- It MUST contain bullet points for any code or technical information shared
- It MUST contain a section of key insights gained
- It MUST format the summary in the third person

Example format:

## Conversation Summary
* Topic 1: Key information
* Topic 2: Key information

## Tools Executed
* Tool X: Result Y`

/**
 * Adjust a split point forward to avoid breaking tool use/result pairs.
 *
 * Walks the split point forward until the message at that position is neither
 * an orphaned toolResult nor a toolUse without an immediately following toolResult.
 *
 * @throws If no valid split point can be found (walked past all messages)
 */
export function adjustSplitPointForToolPairs(messages: Message[], splitPoint: number): number {
  if (splitPoint >= messages.length) {
    return splitPoint
  }

  while (splitPoint < messages.length) {
    const message = messages[splitPoint]!

    const hasToolResult = message.content.some((block) => block.type === 'toolResultBlock')
    if (hasToolResult) {
      splitPoint++
      continue
    }

    const hasToolUse = message.content.some((block) => block.type === 'toolUseBlock')
    if (hasToolUse) {
      const nextMessage = messages[splitPoint + 1]
      const nextHasToolResult = nextMessage?.content.some((block) => block.type === 'toolResultBlock')
      if (!nextHasToolResult) {
        splitPoint++
        continue
      }
    }

    break
  }

  if (splitPoint >= messages.length) {
    throw new Error('Unable to find valid split point for summarization')
  }

  return splitPoint
}

/**
 * Find a valid trim point for truncation starting at `startIndex`.
 *
 * A valid trim point must:
 * 1. Be a user message (required by some models)
 * 2. Not be an orphaned toolResult
 * 3. Not be a toolUse unless its toolResult immediately follows
 *
 * @returns The valid trim index, or `messages.length` if none found
 */
export function findValidTrimPoint(messages: Message[], startIndex: number): number {
  let trimIndex = startIndex

  while (trimIndex < messages.length) {
    const message = messages[trimIndex]
    if (!message) break

    if (message.role !== 'user') {
      trimIndex++
      continue
    }

    const hasToolResult = message.content.some((block) => block.type === 'toolResultBlock')
    if (hasToolResult) {
      trimIndex++
      continue
    }

    const hasToolUse = message.content.some((block) => block.type === 'toolUseBlock')
    if (hasToolUse) {
      const nextMessage = messages[trimIndex + 1]
      const nextHasToolResult = nextMessage && nextMessage.content.some((block) => block.type === 'toolResultBlock')
      if (!nextHasToolResult) {
        trimIndex++
        continue
      }
    }

    break
  }

  return trimIndex
}

/**
 * Generate a summary of the provided messages by calling the model.
 *
 * @returns A user-role message containing the model-generated summary
 * @throws If the model fails to produce a response
 */
export async function generateSummary(
  messagesToSummarize: Message[],
  model: Model,
  systemPrompt?: string
): Promise<Message> {
  const summarizationMessages = [
    ...messagesToSummarize,
    new Message({
      role: 'user',
      content: [new TextBlock('Please summarize this conversation.')],
    }),
  ]

  const stream = model.streamAggregated(summarizationMessages, {
    systemPrompt: systemPrompt ?? DEFAULT_SUMMARIZATION_PROMPT,
  })

  let result: Awaited<ReturnType<typeof stream.next>> | undefined
  for (;;) {
    result = await stream.next()
    if (result.done) break
  }

  if (!result?.done || !result.value) {
    throw new Error('Failed to generate summary: no response from model')
  }

  return new Message({
    role: 'user',
    content: result.value.message.content,
  })
}

export type MessageTypeFilter = 'tools' | 'messages' | 'all'

/**
 * Returns true if the message matches the given type filter.
 * - 'tools': message contains at least one toolUseBlock or toolResultBlock
 * - 'messages': message contains no toolUseBlock or toolResultBlock
 * - 'all': always matches
 */
export function matchesMessageType(message: Message, filter: MessageTypeFilter): boolean {
  if (filter === 'all') return true
  const hasTool = message.content.some((b) => b.type === 'toolUseBlock' || b.type === 'toolResultBlock')
  if (filter === 'tools') return hasTool
  if (filter === 'messages') return !hasTool
  return false
}
